File: frontend/lib/sentry.ts
"use client";

/**
 * Frontend Sentry initialisation.
 *
 * Call `initSentry()` once at application startup (e.g. in a root layout or
 * _app component). Initialisation is skipped when NEXT_PUBLIC_SENTRY_DSN is
 * unset, so local development and tests are unaffected by default.
 *
 * PII scrubbing applied via beforeSend:
 *   - Email addresses in exception messages are replaced with "[email]".
 *   - Request body, cookies, and the Authorization/Cookie headers are stripped.
 */

import * as Sentry from "@sentry/nextjs";

const EMAIL_RE = /[a-zA-Z0-9_.+\-]+@[a-zA-Z0-9\-]+\.[a-zA-Z0-9\-.]+/g;

function scrubEmails(value: string): string {
  return value.replace(EMAIL_RE, "[email]");
}

function beforeSend(event: Sentry.ErrorEvent): Sentry.ErrorEvent | null {
  // Scrub emails from exception messages
  for (const exc of event.exception?.values ?? []) {
    if (exc.value) {
      exc.value = scrubEmails(exc.value);
    }
  }

  // Strip request body and sensitive headers
  if (event.request) {
    delete (event.request as Record<string, unknown>).data;
    delete (event.request as Record<string, unknown>).cookies;
    const headers = event.request.headers as Record<string, string> | undefined;
    if (headers) {
      delete headers["Authorization"];
      delete headers["authorization"];
      delete headers["Cookie"];
      delete headers["cookie"];
    }
  }

  return event;
}

export function initSentry(): void {
  const dsn = process.env.NEXT_PUBLIC_SENTRY_DSN;
  if (!dsn) return;

  const rawRate = process.env.NEXT_PUBLIC_SENTRY_TRACES_SAMPLE_RATE;
  const tracesSampleRate = rawRate !== undefined ? parseFloat(rawRate) : 0.1;

  Sentry.init({
    dsn,
    environment: process.env.NEXT_PUBLIC_SENTRY_ENVIRONMENT ?? "development",
    tracesSampleRate,
    beforeSend,
    sendDefaultPii: false,
  });
}

export { Sentry };

File: backend/app/services/sentry.py
"""Sentry error-reporting integration.

PII scrubbing rules applied via `before_send`:
- Email addresses in exception values are replaced with "[email]".
- Request body (`data`), cookies, and the `Authorization`/`Cookie` headers
  are stripped so message bodies and tokens never reach Sentry.
"""

from __future__ import annotations

import logging
import re
from typing import Any

import sentry_sdk
from sentry_sdk.integrations.fastapi import FastApiIntegration
from sentry_sdk.integrations.starlette import StarletteIntegration
from sentry_sdk.types import Event, Hint

from app.config import get_settings

logger = logging.getLogger(__name__)

_EMAIL_RE = re.compile(r"[a-zA-Z0-9_.+\-]+@[a-zA-Z0-9\-]+\.[a-zA-Z0-9\-.]+")


def _before_send(event: Event, hint: Hint) -> Event | None:
    # Scrub email addresses that may appear in exception messages
    for exc in (event.get("exception") or {}).get("values", []):
        if exc.get("value"):
            exc["value"] = _EMAIL_RE.sub("[email]", exc["value"])

    # Strip request body and sensitive headers
    req: dict[str, Any] = event.get("request") or {}
    req.pop("data", None)
    req.pop("cookies", None)
    headers: dict[str, Any] = req.get("headers") or {}
    for key in list(headers):
        if key.lower() in {"authorization", "cookie"}:
            del headers[key]

    return event


def init_sentry() -> None:
    settings = get_settings()
    if not settings.sentry_dsn:
        logger.info("Sentry DSN not configured — skipping initialisation")
        return

    sentry_sdk.init(
        dsn=settings.sentry_dsn,
        environment=settings.sentry_environment,
        traces_sample_rate=settings.sentry_traces_sample_rate,
        integrations=[
            StarletteIntegration(),
            FastApiIntegration(),
        ],
        before_send=_before_send,
        send_default_pii=False,
    )
    logger.info("Sentry initialised for environment '%s'", settings.sentry_environment)
